drivers/timer/riscv_machine_timer.c

```diff
@@ -1,117 +1,135 @@
+#define MIN_DELAY 1000
 
-#include "legacy_api.h"
+#define TICKLESS (IS_ENABLED(CONFIG_TICKLESS_KERNEL) &&		\
+		  !IS_ENABLED(CONFIG_QEMU_TICKLESS_WORKAROUND))
 
-typedef struct {
-	u32_t val_low;
-	u32_t val_high;
-} riscv_machine_timer_t;
+static struct k_spinlock lock;
+static u64_t last_count;
 
-static volatile riscv_machine_timer_t *mtime =
-	(riscv_machine_timer_t *)RISCV_MTIME_BASE;
-static volatile riscv_machine_timer_t *mtimecmp =
-	(riscv_machine_timer_t *)RISCV_MTIMECMP_BASE;
-
-/*
- * The RISCV machine-mode timer is a one shot timer that needs to be rearm upon
- * every interrupt. Timer clock is a 64-bits ART.
- * To arm timer, we need to read the RTC value and update the
- * timer compare register by the RTC value + time interval we want timer
- * to interrupt.
- */
-static ALWAYS_INLINE void riscv_machine_rearm_timer(void)
+static void set_mtimecmp(u64_t time)
 {
-	u64_t rtc;
+	volatile u32_t *r = (u32_t *)RISCV_MTIMECMP_BASE;
 
-	/*
-	 * Disable timer interrupt while rearming the timer
-	 * to avoid generation of interrupts while setting
-	 * the mtimecmp->val_low register.
+	/* Per spec, the RISC-V MTIME/MTIMECMP registers are 64 bit,
+	 * but are NOT internally latched for multiword transfers.  So
+	 * we have to be careful about sequencing to avoid triggering
+	 * spurious interrupts: always set the high word to a max
+	 * value first.
 	 */
-	irq_disable(RISCV_MACHINE_TIMER_IRQ);
-
-	/*
-	 * Following machine-mode timer implementation in QEMU, the actual
-	 * RTC read is performed when reading low timer value register.
-	 * Reading high timer value just reads the most significant 32-bits
-	 * of a cache value, obtained from a previous read to the low
-	 * timer value register. Hence, always read timer->val_low first.
-	 * This also works for other implementations.
-	 */
-	rtc = mtime->val_low;
-	rtc |= ((u64_t)mtime->val_high << 32);
+	r[1] = 0xffffffff;
+	r[0] = (u32_t)time;
+	r[1] = (u32_t)(time >> 32);
+}
 
-	/*
-	 * Rearm timer to generate an interrupt after
-	 * sys_clock_hw_cycles_per_tick()
-	 */
-	rtc += sys_clock_hw_cycles_per_tick();
-	mtimecmp->val_low = (u32_t)(rtc & 0xffffffff);
-	mtimecmp->val_high = (u32_t)((rtc >> 32) & 0xffffffff);
+static u64_t mtime(void)
+{
+	volatile u32_t *r = (u32_t *)RISCV_MTIME_BASE;
+	u32_t lo, hi;
 
-	/* Enable timer interrupt */
-	irq_enable(RISCV_MACHINE_TIMER_IRQ);
+	/* Likewise, must guard against rollover when reading */
+	do {
+		hi = r[1];
+		lo = r[0];
+	} while (r[1] != hi);
+
+	return (((u64_t)hi) << 32) | lo;
 }
 
-static void riscv_machine_timer_irq_handler(void *unused)
+static void timer_isr(void *arg)
 {
-	ARG_UNUSED(unused);
-#ifdef CONFIG_EXECUTION_BENCHMARKING
-	extern void read_timer_start_of_tick_handler(void);
-	read_timer_start_of_tick_handler();
-#endif
+	ARG_UNUSED(arg);
 
-	z_clock_announce(1);
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	u32_t now = mtime();
+	u32_t dticks = ((u32_t)now - (u32_t)last_count) / CYC_PER_TICK;
```

Comment:
I think these u32_t's are causing overflow bugs when I run some tests on hardware for an extended period of time. I haven't sat down and really thought through the math, but I get weird timer behavior coinciding with mtime values larger than 32 bits. Changing these variables to u64_t's fixes it.

Reply:
Huh.  It certainly looks right.  How long an extended period are we talking?  Can you rerun the test with both variants and an _ASSERT to check that they are equivalent?  Does it work if you do the "(now-last_count)/CYC_PER_TICK" in 64 bit but then cast for dticks?

Compiler bugs happen too -- do you have a more recent toolchain to try than the SDK?

Reply:
On the platform I'm testing on, the machine timer clock is driven at 12.5 MHz, and the misbehavior happens consistently after just over 5 minutes and 30 seconds, the same time that `mtime` rolls over to 33 bits.

I'll try the tests you suggest and see what I can figure out.

Reply:
Setting an __ASSERT does detect a discrepancy between the argument to `set_mtimecmp()` when the mtime rollover happens. The deciding factor looks like it might be when truncating `now` to 32-bits.

The code that works is
```
u64_t now = mtime();
u32_t dticks = (u32_t)((now - last_count) / CYC_PER_TICK);
```

I was building on SDK 0.9.3. Updating to 0.9.4-rc3 still exhibits the problem, so at least it's not  a compiler bug solved between those two releases.